kernel/posix/pthread.c

```diff
@@ -0,0 +1,550 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <atomic.h>
+#include "ksched.h"
+#include "wait_q.h"
+
+const pthread_attr_t init_pthread_attrs = {
+	.priority = K_LOWEST_APPLICATION_THREAD_PRIO,
+	.stack = NULL,
+	.stacksize = 0,
+	.flags = PTHREAD_INIT_FLAGS,
+	.delayedstart = K_NO_WAIT,
+	.schedpolicy = SCHED_RR,
+	.detachstate = PTHREAD_CREATE_JOINABLE,
+	.initialized = true,
+};
+
+/* Pthread object for number of threads */
+static struct posix_thread pthread[CONFIG_MAX_PTHREAD_COUNT];
+
+static bool is_posix_prio_valid(u32_t priority, int policy)
+{
+	if (priority >= sched_get_priority_min(policy) &&
+	    priority <= sched_get_priority_max(policy)) {
+		return true;
+	}
+
+	return false;
+}
+
+static u32_t zephyr_to_posix_priority(s32_t z_prio, int *policy)
+{
+	u32_t prio;
+
+	if (z_prio < 0) {
+		*policy = SCHED_FIFO;
+		prio = -1 * (z_prio + 1);
+	} else {
+		*policy = SCHED_RR;
+		prio = (CONFIG_NUM_PREEMPT_PRIORITIES - z_prio);
+	}
+
+	return prio;
+}
+
+static s32_t posix_to_zephyr_priority(u32_t priority, int policy)
+{
+	s32_t prio;
+
+	if (policy == SCHED_FIFO) {
+		/* Zephyr COOP priority starts from -1 */
+		prio =  -1 * (priority + 1);
+	} else {
+		prio = (CONFIG_NUM_PREEMPT_PRIORITIES - priority);
+	}
+
+	return prio;
+}
+
+/**
+ * @brief Set scheduling parameter attributes in thread attributes object.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_attr_setschedparam(pthread_attr_t *attr,
+			       const struct sched_param *schedparam)
+{
+	int priority = schedparam->priority;
+
+	if (!attr || !attr->initialized) {
+		errno = EINVAL;
```

Comment:
pthread functions do not set errno, they return the error code directly.

Reply:
initially I was thinking of return -1 and send errno on Error, but looks like to be strictly compliant with POSIX, i will return -ERROR_CODE directly. I will modify this for all functions.

Reply:
It's ERROR_CODE; not -ERROR_CODE. Yes, I know it's confusing and inconsistent with other APIs in POSIX, but that's how it's defined.